cli/cmd/top.go

```diff
@@ -0,0 +1,378 @@
+	if success {
+		switch eos := req.rspEnd.GetEos().GetEnd().(type) {
+		case *pb.Eos_GrpcStatusCode:
+			success = eos.GrpcStatusCode == 0
+
+		case *pb.Eos_ResetErrorCode:
+			success = false
+		}
+	}
+
+	found := false
+	for i, row := range *table {
+		if row.by == by && row.source == source && row.destination == destination {
+			(*table)[i].count++
+			if latency.Nanos < row.best.Nanos {
+				(*table)[i].best = latency
+			}
+			if latency.Nanos > row.worst.Nanos {
+				(*table)[i].worst = latency
+			}
+			(*table)[i].last = latency
+			if success {
+				(*table)[i].successes++
+			} else {
+				(*table)[i].failures++
+			}
+			found = true
+		}
+	}
+
+	if !found {
+		successes := 0
+		failures := 0
+		if success {
+			successes++
+		} else {
+			failures++
+		}
+		row := tableRow{
+			by:          by,
+			source:      source,
+			destination: destination,
+			count:       1,
+			best:        latency,
+			worst:       latency,
+			last:        latency,
+			successes:   successes,
+			failures:    failures,
+		}
+		*table = append(*table, row)
+	}
+}
+
+func stripPort(address string) string {
+	return strings.Split(address, ":")[0]
+}
+
+func renderHeaders() {
+	tbprint(0, 0, "(press q to quit)")
+	x := 0
+	for i, header := range columnNames {
+		width := columnWidths[i]
+		padded := fmt.Sprintf("%-"+strconv.Itoa(width)+"s ", header)
+		tbprintBold(x, 2, padded)
+		x += width + 1
+	}
+}
+
+func renderTableBody(table []tableRow) {
+	sort.SliceStable(table, func(i, j int) bool {
+		return table[i].count > table[j].count
+	})
+	for i, row := range table {
+		x := 0
+		tbprint(x, i+headerHeight, row.source)
+		x += columnWidths[0] + 1
+		tbprint(x, i+headerHeight, row.destination)
+		x += columnWidths[1] + 1
+		tbprint(x, i+headerHeight, row.by)
+		x += columnWidths[2] + 1
+		tbprint(x, i+headerHeight, strconv.Itoa(row.count))
+		x += columnWidths[3] + 1
+		tbprint(x, i+headerHeight, formatDuration(row.best))
+		x += columnWidths[4] + 1
+		tbprint(x, i+headerHeight, formatDuration(row.worst))
+		x += columnWidths[5] + 1
+		tbprint(x, i+headerHeight, formatDuration(row.last))
+		x += columnWidths[6] + 1
+		successRate := fmt.Sprintf("%.2f%%", 100.0*float32(row.successes)/float32(row.successes+row.failures))
+		tbprint(x, i+headerHeight, successRate)
+	}
+}
+
+func tbprint(x, y int, msg string) {
+	for _, c := range msg {
+		termbox.SetCell(x, y, c, termbox.ColorDefault, termbox.ColorDefault)
+		x += runewidth.RuneWidth(c)
+	}
+}
+
+func tbprintBold(x, y int, msg string) {
+	for _, c := range msg {
+		termbox.SetCell(x, y, c, termbox.AttrBold, termbox.ColorDefault)
+		x += runewidth.RuneWidth(c)
+	}
+}
+
+func formatDuration(d duration.Duration) string {
+	if d.Nanos < 1000000 {
+		micros := d.Nanos / 1000
+		return fmt.Sprintf("%dµs", micros)
+	}
+	if d.Nanos < 1000000000 {
+		millis := d.Nanos / 1000000
+		return fmt.Sprintf("%dms", millis)
+	}
+	secs := d.Nanos / 1000000000
+	return fmt.Sprintf("%ds", secs)
+}
```

Comment:
I find `time.Duration` way easier to work with than protobuf durations. If this method accepted `time.Duration` as a parameter, then I'd write it as:

```
func formatDuration(d time.Duration) string {
	if d < time.Millisecond {
		return d.Round(time.Microsecond).String()
	}
	if d < time.Second {
		return d.Round(time.Millisecond).String()
	}
	return d.Round(time.Second).String()
}
```

To do that, I'd convert the protobuf duration to a `time.Duration` as soon as it's received (using [this func](https://github.com/golang/protobuf/blob/master/ptypes/duration.go#L76), and then use `time.Duration` everywhere).